tools/saftw.py

```diff
@@ -0,0 +1,310 @@
+
+def main():
+    co = cmdopt()
+    try:
+        opts, args = getopt.getopt(sys.argv[1:], "no:mp:dJt:r:Rj:l:w")
+    except getopt.GetoptError as e:
+        usage()
+    for o, a in opts:
+        if o == "-m":
+            co.rulematch = 1
+        if o == "-d":
+            co.debug = 1
+        if o == "-t":
+            co.timeout = int(a)
+        if o == "-r":
+            co.rule_path = a
+        if o == "-J":
+            co.replayjournal = 1
+        if o == "-j":
+            co.journal = a
+        if o == "-R":
+            co.recurse = 1
+        if o == "-p":
+            co.plugin = a
+        if o == "-o":
+            parse_overload(a, co)
+        if o == "-l":
+            co.logfile = a
+        if o == "-n":
+            co.nocache = 1
+    if co.plugin == None:
+        usage()
+    try:
+        ob = waf_plugin()
+        waf_plug = ob.load(co.plugin)
+    except Exception as e:
+        print e
+        print "Invalid plugin specifed"
+        sys.exit(1)
+    co.plugin = waf_plug
+    if co.rule_path == None and co.replayjournal == 0:
+        usage()
+    if co.journal == None:
+        usage()
+    if co.replayjournal == 1:
+        if co.logfile == None:
+            print "must specify logfile"
+            sys.exit(1)
+        replay_journal(co)
+        return 0
+    try:
+        jrnl = journal(co.journal)
+    except IOError as e:
+        print "failed to initialize journal: {}".format(e.strerror)
+        sys.exit(1)
+    testfiles = util.get_rulesets(co.rule_path, co.recurse)
+    for tf in testfiles:
+        for test in tf.tests:
+            now = time.time()
+            logtag = waf_plug.get_log_tag(now, test.test_title)
+            runner = testrunner.TestRunner()
+            for stage in test.stages:
+                do_overload(stage, co)
+                odict = stage.output.output_dict
+                invert = 0
+                if "no_log_contains" in odict.keys():
+                    invert = 1
+                headers = stage.input.headers
+                if co.nocache == 1:
+                    if not "Cache-Control" in headers.keys():
+                        stage.input.headers["Cache-Control"] = "no-cache, no-store, must-revalidate"
+                if not "X-Request-Id" in headers.keys():
+                    stage.input.headers["X-Request-Id"] = logtag
+                try:
+                    hua = http.HttpUA()
+                except:
+                    print "failed to initialize UA object"
+                    sys.exit(1)
+                try:
+                    runner.run_stage(stage, None, hua)
+                except Exception as e:
+                    print e
+                    pass
+                if co.debug == 1:
+                    print "Executed test {}".format(test.test_title)
+                jrnl.add_entry(now, test.test_title, logtag, invert, hua.response_object)
+    return 0
+    
+if __name__ == "__main__":
+    main()
+0
```

Comment:
Not sure what this 0 is doing here (or if was part of the original code) but we could probably get rid of it